### PR TITLE
Remove crossing hellos from Channel

### LINKED ISSUE #157: Remove "crossing hellos" logic from channel handshake
[Quote](https://github.com/cosmos/ibc/issues/628):

> 1. Deprecate Version from the ChannelOpenTry message. If the relayer is going to specify a particular version, it should be the INIT relayer choice that determines the ultimate version; not a combination of INIT relayer choice and TRY relayer choice. Note: we will not remove the field as this is a breaking change, we will simply ignore the inputted value.
> 2. Change OnChanOpenTry to return version string, err error. Core IBC will then set the channel struct with the version chosen by OnChanOpenTry

### PR DESCRIPTION
**Pull request by plafer**
Closes: #157

Note that #184 was opened to resolve the current awkward and inconsistent strategy to return values from our handlers/callbacks.

## Description

Still todo:
- [x] cleanup `channel_dispatch()` and `channel_callback()`
    - it currently works but is messy
- [x] Remove `ModuleOutputBuilder` argument from all channel callbacks
- [x] Remove crossing hellos logic from `chan_open_try::process()`
- [x] Deprecate `MsgChannelOpenTry::previous_channel_id` field
- [x] Modify ICS-20 implementation to work with new handshake protocol

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests.
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).
- [x] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/src/mock/context.rs [hunk 1]
```diff
@@ -1605,18 +1606,40 @@ mod tests {
...
                 _port_id: &PortId,
                 _channel_id: &ChannelId,
                 _counterparty: &Counterparty,
-                _version: &Version,
                 counterparty_version: &Version,
-            ) -> Result<Version, Error> {
-                Ok(counterparty_version.clone())
+            ) -> Result<(ModuleExtras, Version), Error> {
+                Ok((
+                    ModuleExtras {
```
Comment (romac) [inline crates/ibc/src/mock/context.rs]: Same here

File: crates/ibc/src/mock/context.rs [hunk 2]
```diff
@@ -1605,18 +1606,40 @@ mod tests {
...
+                &mut self,
+                _order: Order,
+                _connection_hops: &[ConnectionId],
+                _port_id: &PortId,
+                _channel_id: &ChannelId,
+                _counterparty: &Counterparty,
+                version: &Version,
+            ) -> Result<(ModuleExtras, Version), Error> {
+                Ok((
+                    ModuleExtras {
```
Comment (romac) [inline crates/ibc/src/mock/context.rs]: We could use a `Default` instance here, or an `empty()` constructor.

File: crates/ibc/src/core/ics26_routing/context.rs [hunk 3]
```diff
@@ -101,65 +102,68 @@ pub trait Module: Send + Sync + AsAnyMut {
...
 
     fn on_chan_close_confirm(
         &mut self,
-        _output: &mut ModuleOutputBuilder,
         _port_id: &PortId,
         _channel_id: &ChannelId,
-    ) -> Result<(), Error> {
-        Ok(())
+    ) -> Result<ModuleExtras, Error> {
+        Ok(ModuleExtras {
```
Comment (romac) [inline crates/ibc/src/core/ics26_routing/context.rs]: Same here

File: crates/ibc/src/core/ics26_routing/context.rs [hunk 4]
```diff
@@ -101,65 +102,68 @@ pub trait Module: Send + Sync + AsAnyMut {
...
 
     fn on_chan_close_init(
         &mut self,
-        _output: &mut ModuleOutputBuilder,
         _port_id: &PortId,
         _channel_id: &ChannelId,
-    ) -> Result<(), Error> {
-        Ok(())
+    ) -> Result<ModuleExtras, Error> {
+        Ok(ModuleExtras {
```
Comment (romac) [inline crates/ibc/src/core/ics26_routing/context.rs]: Same here

File: crates/ibc/src/core/ics26_routing/context.rs [hunk 5]
```diff
@@ -101,65 +102,68 @@ pub trait Module: Send + Sync + AsAnyMut {
...
 
     fn on_chan_open_confirm(
         &mut self,
-        _output: &mut ModuleOutputBuilder,
         _port_id: &PortId,
         _channel_id: &ChannelId,
-    ) -> Result<(), Error> {
-        Ok(())
+    ) -> Result<ModuleExtras, Error> {
+        Ok(ModuleExtras {
```
Comment (romac) [inline crates/ibc/src/core/ics26_routing/context.rs]: Same here

File: crates/ibc/src/core/ics26_routing/context.rs [hunk 6]
```diff
@@ -101,65 +102,68 @@ pub trait Module: Send + Sync + AsAnyMut {
...
         &mut self,
-        _output: &mut ModuleOutputBuilder,
         _port_id: &PortId,
         _channel_id: &ChannelId,
         _counterparty_version: &Version,
-    ) -> Result<(), Error> {
-        Ok(())
+    ) -> Result<ModuleExtras, Error> {
+        Ok(ModuleExtras {
+            events: Vec::new(),
```
Comment (romac) [inline crates/ibc/src/core/ics26_routing/context.rs]: Same here

File: crates/ibc/src/mock/context.rs [hunk 7]
```diff
@@ -1640,18 +1663,40 @@ mod tests {
...
```
Comment (romac) [inline crates/ibc/src/mock/context.rs]: Same here

File: crates/ibc/src/mock/context.rs [hunk 8]
```diff
@@ -1640,18 +1663,40 @@ mod tests {
...
                 _port_id: &PortId,
                 _channel_id: &ChannelId,
                 _counterparty: &Counterparty,
-                _version: &Version,
                 counterparty_version: &Version,
-            ) -> Result<Version, Error> {
-                Ok(counterparty_version.clone())
+            ) -> Result<(ModuleExtras, Version), Error> {
+                Ok((
+                    ModuleExtras {
```
Comment (romac) [inline crates/ibc/src/mock/context.rs]: Same here

**Review by romac — APPROVED** [review]
Left a nitpick but looks good otherwise!

File: crates/ibc/src/mock/context.rs (hunk 2 above)
Comment (plafer) [reply crates/ibc/src/mock/context.rs]: [516f520](https://github.com/cosmos/ibc-rs/pull/181/commits/516f520971330245d14a77965924cbd88b341640) (fixes all)

**Pull request merged**